File: 63. Unique Paths II.py
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        m = len(obstacleGrid)
        n = len(obstacleGrid[0])
        
        if obstacleGrid[0][0] == 1 or obstacleGrid[-1][-1] == 1:
            return 0
        
        result = [[0 for _ in range(n)] for _ in range(m)]
        result[0][0] = 1
        for i in range(1, n):
            result[0][i] = result[0][i-1] if obstacleGrid[0][i] == 0 else 0
        for i in range(1, m):
            result[i][0] = result[i-1][0] if obstacleGrid[i][0] == 0 else 0
        for i in range(1,m):
            for j in range(1,n):
                if obstacleGrid[i][j] == 0:
                    result[i][j] = result[i-1][j] + result[i][j-1]
        return result[-1][-1]

File: 63. Unique Paths II.java
class Solution {
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        if (obstacleGrid==null || obstacleGrid.length==0 || obstacleGrid[0][0]==1) {
            return 0;
        }
        int m = obstacleGrid[0].length;
        int n = obstacleGrid.length;
        int[] dp = new int[m];
        for (int i=0; i<n; i++) {
            for (int j=0; j<m; j++) {
                if (i==0) {
                    dp[j] = obstacleGrid[i][j] == 1 || (j>0 && dp[j-1]==0) ? 0 : 1;
                } else {
                    if (obstacleGrid[i][j] == 1) {
                        dp[j] = 0;
                    } else {
                        if (j==0) {
                            dp[j] = dp[j] == 0 ? 0 : 1;
                        } else {
                            dp[j] += dp[j-1];
                        }
                    }
                }
            }
        }
        return dp[m-1];
    }
}